gpii/node_modules/gpii-oauth2/gpii-oauth2-authz-server/src/AuthorizationService.js

```diff
@@ -100,125 +100,538 @@ var fluid = fluid || require("infusion");
+    // {
+    //     id: (String)
+    // }
+    // The returned structure may contain more fields if the promise is returned by dataStore.addAuthDecision()
+    gpii.oauth2.authorizationService.doAdd = function (dataStore, codeGenerator, record) {
+        var promiseTogo = fluid.promise();
+        if (record.gpiiToken && record.client) {
+            var clientId = record.client.id;
+            var redirectUri = record.client.redirectUri;
+            // Check to see if we have an existing authorization
+            var authDecisionPromise = dataStore.findAuthDecision(record.inputArgs.gpiiToken, clientId, redirectUri);
+            authDecisionPromise.then(function (authDecision) {
+                if (!authDecision) {
+                    // If not, add one
+                    var accessToken = codeGenerator.generateAccessToken();
+                    var addAuthDecisionPromise = dataStore.addAuthDecision({
+                        gpiiToken: record.inputArgs.gpiiToken,
+                        clientId: clientId,
+                        redirectUri: redirectUri,
+                        accessToken: accessToken,
+                        selectedPreferences: record.inputArgs.selectedPreferences,
+                        revoked: false
+                    });
+                    fluid.promise.follow(addAuthDecisionPromise, promiseTogo);
+                } else {
+                    // If the auth decision already exists, return its id
+                    promiseTogo.resolve({id: authDecision.id});
+                }
+            }, function (err) {
+                promiseTogo.reject(err);
+            });
         }
+        return promiseTogo;
     };
+    // ==== End of addAuthorization()
 
-    gpii.oauth2.authorizationService.userHasAuthorized = function (dataStore, userId, clientId, redirectUri) {
-        return dataStore.findAuthDecision(userId, clientId, redirectUri) ? true : false;
+    gpii.oauth2.authorizationService.userHasAuthorized = function (dataStore, gpiiToken, clientId, redirectUri) {
+        var promiseTogo = fluid.promise();
+        var authDecisionPromise = dataStore.findAuthDecision(gpiiToken, clientId, redirectUri);
+        authDecisionPromise.then(function (authDecision) {
+            var result = authDecision ? true : false;
+            promiseTogo.resolve(result);
+        }, function (err) {
+            promiseTogo.reject(err);
+        });
+
+        return promiseTogo;
     };
 
     gpii.oauth2.authorizationService.exchangeCodeForAccessToken = function (dataStore, code, clientId, redirectUri) {
-        var auth = dataStore.findAuthByCode(code);
-        // TODO flag an authCode after it is found to make single use
-        if (auth && auth.clientId === clientId && auth.redirectUri === redirectUri) {
-            return auth.accessToken;
+        var promiseTogo = fluid.promise();
+        var authPromise = dataStore.findAuthByCode(code);
+        authPromise.then(function (auth) {
```

Comment:
What happened to this method is typical for "promisification cruft" : P as well as refactoring with fluid.promise.map, extract the callback to a separate function